Support multiple run/dispose cycles for sources

brandonpayton opened the pull request:
Hi @TylorS,

I am encountering a problem using a subject as a signal for `most.until`. It works the first time, but if I try using the same subject later with `until`, `until` never receives a signal because the subject no longer emits events.

The reason appears to be that a source may be run and disposed multiple times in its lifetime, but BasicSubjectSource only expects to be disposed once. This is a PR to fix that issue.

To see how a source may be run and disposed multiple times, you can examine `UpperBound`, which is used to implement `until`, [calling run on the signal](https://github.com/cujojs/most/blob/2a53262f2b316f333ca85566daaa78bf40eeae45/lib/combinator/timeslice.js#L99) (a source) and [later disposing it](https://github.com/cujojs/most/blob/2a53262f2b316f333ca85566daaa78bf40eeae45/lib/combinator/timeslice.js#L113).

Here is a simplified example where I'm triggering the issue:
```javascript
// mouseDown$, mouseMove$, and mouseUp$ are subjects used to relay DOM events
mouseDown$.chain(
  mouseDownEvent => mouseMove$.until(mouseUp$).map(...)
)
.observe(...)
```
This works the first time `mouseUp$` signals. After that, `mouseUp$` has been disposed, which permanently marks the subject inactive and blocks further events from being relayed. When another `mouseDown$` event arrives, `mouseUp$` is passed to `until` and is `run` but never marks itself as active which would allow events to be relayed again.

I don't love my wording in the commit messages or the name of the new test, but that is the best I came up with tonight. Any suggestions for those would be appreciated. :)

TylorS commented:
Hi @brandonpayton Thank you for this! I've been thinking about this for a bit, and I was wondering what you think of the following ideas:

1. Extend your solution to be more like the rest of most.js - this would entail completing a stream when error() is called. In all of most an error **always** ends and disposes of the stream synchronously.

2. Do no real completion and leave it to the developer to handle those type of things. Or in other words, a Subject would always just be a pass through. This would potentially allow one to call `.error()` or `.complete()` multiple times.

brandonpayton commented:
Hi @TylorS, handling the errors the same way as most sounds great to me. 

Regarding no real completion, I need to review how most supports error recovery this evening to offer a better response. I have a feeling completion should not be forced when `subject.error` is called.

What is your thinking around allowing multiple completions? It's an interesting idea I hadn't considered. My gut feel is that reusing a single subject through multiple completions may lead to hard to find bugs.

TylorS commented:
Yeah, I think option 1 is the safest, and the closest to the most philosophy of doing things, ignoring the fact that Subjects are imperative. I'm mostly in favor of option 1.

Option 2 is interesting. It makes Subjects more powerful, but also more likely to cause bugs like you mentioned. I'm not sure everyone sees Subjects as a last resort in the same way I do, where for me if you're using a Subject, I personally think you should understand the inherent "risk" you're taking and that there are really no safeguards from keeping you from doing something silly.

brandonpayton commented:
cool. I'll take a look at option 1.

Regarding option 2, I tend to view them as a last resort as well but likely have more to learn about why. Mostly, I think they are too tempting to use for people still in the habit of thinking imperatively. If you have other thoughts here, I'm interested in learning more. But even as a last resort and acknowledging the risk, I think it is better to limit risk if you can do so reasonably.

brandonpayton commented:
And maybe I'm overestimating how much risk we can limit. :)

brandonpayton commented:
Hi @TylorS, after looking at `most` and `@most/hold`, I've made the following changes to line up with `most` behavior:
1. Updated BasicSubjectSource to allow new observers after ending by completion or error. I also added some simple tests for this.
2. Call `_dispose` for completion and error. Effectively, this is the same as we had before, but the call to `_dispose` makes clear we are disposing on completion and error.
3. `holdSubject()` now only buffers events when it has an observer. This is the behavior I observed with `@most/hold`. And I also observed that an error does nothing to clear the held event, so new observers to a stream through `@most/hold` will first receive the last event that occurred before the error.

This is now closer to incorporating your thought of subjects being pass-throughs where you can call `complete` and `error` multiple times on them. Given `most`'s behavior this seemed most appropriate.

What do you think?

TylorS commented on test/holdSubject.js:
```diff
@@ -25,6 +25,9 @@ describe('holdSubject', () => {
     const stream = holdSubject(3)
 
+    // Add an observer so the stream begins buffering events
+    stream.observe(() => {})
```
Doesn't matter here, but stream.drain() is actually the same thing :)

brandonpayton replied:
Ah, right. Good to remember. I just needed `drain` the other day but hadn't used it prior.

TylorS commented:
Thank you @brandonpayton LGTM

TylorS commented:
Published in v4.1.2
